### ISSUE TITLE
SDWebImagePrefetcher iOS9

### ISSUE THREAD

**Issue by standers67** [rs/SDWebImage#1366]
I am unable to use the SDWebImagePrefetcher to prefetch images. I am using the iOS 9.1 SDK and have tried to setup some very basic code to preload a few images. Adding more than 2 URLs to the array, makes the code hang.

// The following code works...
NSMutableArray *imageURLs = [NSMutableArray arrayWithCapacity:10];
[imageURLs addObject:[NSURL URLWithString:@"http://site.com/images/41/65/53/3696931799c.png"]];
[imageURLs addObject:[NSURL URLWithString:@"http://site.com/images/41/65/53/3696922399c.png"]];
[[SDWebImagePrefetcher sharedImagePrefetcher] prefetchURLs:imageURLs];

// The following code with one more url does not... (app hangs)
NSMutableArray *imageURLs = [NSMutableArray arrayWithCapacity:10];
[imageURLs addObject:[NSURL URLWithString:@"http://site.com/images/41/65/53/3696931799c.png"]];
[imageURLs addObject:[NSURL URLWithString:@"http://site.com/images/41/65/53/3696922399c.png"]];
[imageURLs addObject:[NSURL URLWithString:@"http://site.com/images/41/65/53/3476922399c.png"]];
[[SDWebImagePrefetcher sharedImagePrefetcher] prefetchURLs:imageURLs];

I am using the latest release of the SDWebImage library.

**Comment by mythodeia** [comment]
@standers67 so the only difference is 1 url that makes the app hang? can you try it in the demo project and report back?

**Comment by standers67** [comment]
Thanks for responding so quickly. 

Just tried to do the same thing in the demo project and the behaviour is the same. I try to prefetch images from the _objects array.

Two items in the _objects array works fine, if I add another one - the app hangs..
I added the call to the sharedImagePrefetcher just after two other statements in the demo app:

[SDWebImageManager.sharedManager.imageDownloader setValue:@"SDWebImage Demo" forHTTPHeaderField:@"AppName"];
    SDWebImageManager.sharedManager.imageDownloader.executionOrder = SDWebImageDownloaderLIFOExecutionOrder;
[[SDWebImagePrefetcher sharedImagePrefetcher] prefetchURLs:_objects];

Are there any special conciderations when it comes to when the prefetchURLs method should be called?

**Comment by standers67** [comment]
Some more input, this seems to be the code location where it goes wrong:

SDWebImagePrefetcher.m:128
dispatch_apply(maxNumberOfImages/self.maxConcurrentDownloads, self.prefetcherQueue, ^(size_t index) {
            size_t i = index * self.maxConcurrentDownloads;
            size_t stop = i + self.maxConcurrentDownloads;
            do {
                [weakSelf startPrefetchingAtIndex:i++];
            } while (i < stop);
        });

**Comment by mythodeia** [comment]
replace dispatch_apply line with this one:
```
dispatch_apply(maxNumberOfImages/self.maxConcurrentDownloads, dispatch_get_global_queue(self.prefetcherQueue, 0), ^(size_t index) {
```
the last commit that addressed this apparently broke it

@bpoplauschi we should revert this https://github.com/rs/SDWebImage/commit/92e38160d60f704a137dfd72c7bea0ea0474e26b
as it hangs the prefetcher

**Comment by standers67** [comment]
I get a type conversion error when I replace this line:

Should self.prefetcherQueue be type casted?

SDWebImagePrefetcher.m:131:97: Incompatible pointer to integer conversion passing 'dispatch_queue_t' (aka 'NSObject<OS_dispatch_queue> *') to parameter of type 'long'

**Comment by mythodeia** [comment]
no errors here! tested and working fine
```
dispatch_apply(maxNumberOfImages/self.maxConcurrentDownloads, dispatch_get_global_queue(self.prefetcherQueue, 0), ^(size_t index) {
    size_t i = index * self.maxConcurrentDownloads;
    size_t stop = i + self.maxConcurrentDownloads;
    do {
        [weakSelf startPrefetchingAtIndex:i++];
    } while (i < stop);
});
```

**Comment by standers67** [comment]
Yes, just updated the code in my own project. It works - many thanks!

**Comment by eikemeier** [comment]
Guys, you are doing [shotgun debugging](https://en.wikipedia.org/wiki/Shotgun_debugging) here. `dispatch_get_global_queue(self.prefetcherQueue, …` might work by accident, because it's getting a concurrent queue with random priority, but this doesn't make it right. Either revert commit 6406d8e526c3c19878a490aac3056dad5b5faf74 or check why it does `prefetcherQueue = dispatch_get_main_queue()`, but `dispatch_get_global_queue(dispatch_get_main_queue(), …` seems stupid to me, even when it ‘works’.

**Comment by mythodeia** [comment]
@eikemeier any ideas why https://github.com/rs/SDWebImage/commit/92e38160d60f704a137dfd72c7bea0ea0474e26b commit hangs the code?

**Comment by eikemeier** [comment]
Let's see what 6406d8e526c3c19878a490aac3056dad5b5faf74 did:

Before we had

    dispatch_async(self.prefetcherQueue, ^{		
        [self startPrefetchingAtIndex:self.requestedCount];
    });

So, something was queued asynchronously on the main queue.

After, we had

    dispatch_apply(…, ^(size_t index) {
        [weakSelf startPrefetchingAtIndex:i++];
    });

since [dispatch_apply](https://developer.apple.com/library/ios/documentation/Performance/Reference/GCD_libdispatch_Ref/index.html#//apple_ref/doc/uid/TP40008079-CH2-SW5) waits for all iterations of the task block to complete, it will never return when called from the queue it is dispatching on. So, calling this function from the main thread will hang in default configuration.

Probably it was meant to be queued on some asynchronous queue, and the bug with `dispatch_get_global_queue(self.prefetcherQueue, …` masked the error, I have no idea how the code should really work. I would suggest to revert 6406d8e526c3c19878a490aac3056dad5b5faf74 and think about this again.

**Comment by eikemeier** [comment]
Or, if you prefer random changes, change

        _prefetcherQueue = dispatch_get_main_queue();

to

        _prefetcherQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_SOMETHING, 0);

which will ‘work’.

**Comment by mythodeia** [comment]
@eikemeier thanks for the info! 
@bpoplauschi what do you think? revert the commit and make the random changes @eikemeier posted or revert and rethink the whole thing?

**Comment by joshuafeldman** [comment]
This issue should probably get reopened as the latest HEAD of master still spins the main thread when prefetching

**Comment by livyathan** [comment]
I'm also having the same issue (Xcode 7.1.1, SDK iOS 9.1, iPhone 6). I didn't really understood everything explained on this issue, so I cannot offer assistance.

Initially I was using the shared Prefetcher, but after encountering this problem I decided to use a new instance. I'm using this code on the viewDidLoad:

```
 self.prefetcher = [[SDWebImagePrefetcher alloc] init];
 [self.prefetcher setPrefetcherQueue: dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0)];
self.prefetcher.options = SDWebImageContinueInBackground;
```

I also call to cancelPrefetching on the dealloc. This is also helpful since I need to be able to cancel the requests created on the current screen without affecting the request created by the previous screen.

This seems to be working fine, but is this correct? I mean as a temporally workaround, or is there a chance this will hang randomly? Is it recommended to not use the prefetcher for now?

**Comment by robertmryan** [comment]
OMG, I just saw this `SDWebImagePrefetcher` code and while the `dispatch_apply` stuff was well-intended, it's a disaster. Using `dispatch_apply` for this purpose is completely wrong and suggests some very fundamental misunderstanding on the part of that author of that commit. The purpose of `dispatch_apply` is to run a loop computationally intensive synchronous tasks in parallel on multiple threads and waiting for that to finish. But these tasks are not computationally intensive, nor are the synchronous, nor should we ever wait for anything from the main thread (which is where one would likely call this prefetch routine). And then, adding insult to injury, `prefetchQueue` is defaulting to the main thread, which deadlocks. 

In short, you never use `dispatch_apply` for initiating a bunch of asynchronous operations.

As @eikemeier has suggested, this should be reverted.

**Comment by bpoplauschi** [comment]
Guys, sorry for not reviewing the initial change properly, merging it and taking so long to read this thread. I totally agree with @robertmryan and @eikemeier and will revert the entire MR.

**Comment by bpoplauschi** [comment]
Actually, I only reverted the `dispatch_apply` part that was causing issues, as I considered the rest of the PR valuable

**Comment by livyathan** [comment]
There still a problem here, it looks like it will only prefetch the same amount of images defined by maxConcurrentDownloads:

```
 for (NSUInteger i = 0; i < self.maxConcurrentDownloads && self.requestedCount < listCount; i++) {
            [self startPrefetchingAtIndex:i];
        }
```

This part makes sense, but the method startPrefetchingAtIndex never triggers the download of the next image, and it doesn't look like it's being called from somewhere else, so the rest of the array is being ignored.

**Comment by burniejm** [comment]
i have the same issue as @livyathan.. it just downloads 3 images (or whatever maxConcurrentDownloads is) until i call prefetchURLS again.  Am i missing something here?

**Comment by livyathan** [comment]
It was wrong on my side to report this here, this is a new error, but somebody already reported it on issue #1425, and duplicated on #1426 .

**Comment by bpoplauschi** [comment]
Guys, you were right. When I reverted the `dispatch_apply`, I missed the part that was starting a new prefetch after another one finished. It should be fixed by a8fca65. A new release will go out soon with this fix (3.7.5).
